### ISSUE TITLE
Automatic generation of patterns (a la case class)

### ISSUE THREAD

**Issue by gerardbosch** [vavr-io/vavr#2604]
This is between a question and an idea (may have already been raised by others) regarding structural pattern matching.

Do you think it could be viable an annotation (maybe `@CaseClass`?, maybe other name) that when applied on a class could generate the same code that `@Patterns` + `@Unapply` generates? This is without the necessity of explicitly defining the `@Unapply` method and generate it for example from the class `final` fields instead.

As an example:

```java
@CaseClass
@RequiredArgsConstructor    // Lombok
@Getter                     // Lombok
public class Person {
    private final String firstName;
    private final String lastName;
    private final LocalDate birthDate;
}
```

This would alleviate of writing something like the following to actually generate the `$Person()` function that is used as the pattern:

```java
    @Unapply
    public static Tuple3<String, Integer, Address> Person(Person person) {
        return Tuple(
                person.getFirstName(),
                person.getLast(),
                person.getBirthDate()
        );
    }
```

Would that make sense to simplify the structural pattern matching in Vavr? Cheers!

**Comment by rkraneis** [comment]
I believe @danieldietrich wrote somewhere that he actually wants to trim vavr down and retire features that “are about to be implemented” in Java natively. For structural pattern matching Brian Goetz has already sketched a possible future in [Pattern Matching for Java](https://cr.openjdk.java.net/~briangoetz/amber/pattern-match.html).

First signs are also there already:
- [JEP 305: Pattern Matching for instanceof (Preview)](https://openjdk.java.net/jeps/305)
- [JEP 375: Pattern Matching for instanceof (Second Preview)](https://openjdk.java.net/jeps/375)
- [JEP draft: Pattern matching for switch (Preview)](https://openjdk.java.net/jeps/8213076)

But of course no timeline.

**Comment by gerardbosch** [comment]
Yes, it seems Java will adopt structural pattern matching, although it seems to have no clear timeline for that.

On the other side, Vavr already provides a pattern matching implementation that works from Java 8 :) and very helpful for those that are far away from updating to Java 14+.

So maybe polishing it a little more can help for those on "legacy" Java versions :) It was just an idea, just to see if it would be viable or my suggestion is a no sense :)

Cheers!

**Comment by danieldietrich** [comment]
Hi @gerardbosch !
Generally, I'm open for additions that support legacy Java versions.
@rkraneis is right, the Match API isn't my main focus anymore, I plan to move it to a dedicated repository. There will be new releases but as a separate Java module / jar. So the Vavr users can decide if the want to opt-in.
